### ISSUE TITLE
rename forEachObjIndexed?

### ISSUE THREAD

**Issue by KiaraGrouwstra** [ramda/ramda#2091]
So `forEachObjIndexed` works on arrays as well as on objects, I guess just like `mapObjIndexed`, `filter`, `reject`, `partition`, `pluck`, and a bunch of others like the `prop` / `path` / `lens` categories (search for `Struct` [here](https://github.com/types/npm-ramda/blob/master/index.d.ts)).

Now one part that's a bit misleading there is the categories in the docs, which would lead people to believe only one could be used. For the cases of `forEachObjIndexed` and `mapObjIndexed`, however, this goes a bit further, as even the names would lead one to believe the functions could or should only be used for objects.

Now, one idea here would be to just drop the `Obj` from the names, leaving e.g. `forEachIndexed`, although this remains a bit misleading here -- there is also `addIndex` adding a second argument, which differs a bit from these `Indexed` versions, adding numerical indices even in the case of objects as opposed to their keys. This made me wonder, would `forEachKeyed` be more appropriate to clarify the behavior as distinct from the `addIndex` behavior?

I would propose to have things follow the lead of `map`, which has stayed neutral in name.
Perhaps more controversially, I appreciate Lodash's idea of using a more neutrally named category for functions applicable to both -- in their case `Collection`.

For further consistency, could `forEachObjIndexed` and `mapObjIndexed` be derived using a new `addKey` function mirroring (/ replacing / updating, presuming no-one used it on objects) `addIndex`?

(Note: I'm not sure I like this last idea personally -- I'd rather have Ramda allows users to import useful functions directly instead of forcing them to define everything themselves. But the existence of `addIndex` begs the question whether that's preferred here.)

**Comment by CrossEye** [comment]
@tycho01:

There are a number of ideas here.  Is there one you would most want to promote?

The only one that seems a non-starter to me is 

> I appreciate Lodash's idea of using a more neutrally named category for functions applicable to both -- in their case Collection.

`map` is already dedicated to a very general category, that of `Functors`.  And I can't think of a way to associate `forEach` with Functors.  In fact, there's a huge part of me that would rather remove the `forEach` functions altogether, as they don't fit well with what Ramda tries to do.

**Comment by davidchambers** [comment]
> In fact, there's a huge part of me that would rather remove the `forEach` functions altogether, as they don't fit well with what Ramda tries to do.

Let's do it!

**Comment by tycho01** [comment]
> `map` is already dedicated to a very general category, that of `Functors`.

Yeah, where usable the ADTs seem like a fair method of abstraction -- that does feel in line with how Ramda is positioned (relative to Lodash). We might need to find a golden mean between between using them to clarify (-> educating users `map` works on objects) vs. adding to confusion (partly categorizing based on JS types, partly based on ADTs). In that case, perhaps that's a longer-term consideration, potentially adding more value when type support has broadened (e.g. ES6 types). Perhaps the `@category` tags could be reflected in the docs better as well (-> showing [multiple](https://github.com/ramda/ramda/blob/master/src/addIndex.js#L19-L20)), such as to group the likes of `over` with a 'lens' tag as well.

My understanding of the code-base has been a lot of functions use old-fashioned iteration internally, supporting multiple types as a result (object or array, array or string, usually also any array-likes such as Immutable's `List`...). In the [TS definitions](https://github.com/types/npm-ramda/blob/master/index.d.ts), I've largely attempted to find generalized categories to express this as well (arbitrarily naming it `Struct` in the case of object or array-likes, `List` where either array-likes or string sufficed), if based on my tests rather than the (more modest) Ramda docs. In JavaScript terms though, maybe [`Iterable`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) could be a thing?

> In fact, there's a huge part of me that would rather remove the `forEach` functions altogether, as they don't fit well with what Ramda tries to do.

I used them a bit evidently (having made this thread), but having seen the `R.invoke` use-cases in the other thread now, I'd be fine with it if Ramda were to defer its role as "standard functions except point-free" lib, though users might appreciate pointers as to such alternatives. That said, I do think Ramda was adding a bit of value here by supplementing point-free curried alternatives with customization such as by that [`R.addIndex`](https://github.com/ramda/ramda/issues/1487).

> There are a number of ideas here. Is there one you would most want to promote?

On `forEach`, it seems Lodash's [modules](https://www.npmjs.com/package/lodash.pick) for separate packages serve to aid in providing some metric of usage for their different functions. Presuming it carries any weight, it'd be nice if we'd have some statistics on common Ramda usage as well. That said, I'm not sure if many would bother importing separate packages over the whole thing...

On a potential `R.addKey`: I think this could potentially add value for `countBy`, `eqBy`, `maxBy`, `minBy`, and could allow derivation of `mapObjIndexed`, `foreachObjIndexed`, and maybe even `pickBy` (given a simpler variant with unary parameter for that one). With that, `mapObjIndexed` could live on under a name like `mapKeyed`, / `foreachObjIndexed` (if kept) as `foreachKeyed` or the like. If trimming library size is deemed a priority, derived functions like that could be relegated back to cookbooks or the like.

I made a shot at adding an `R.addKey` [here](https://github.com/ramda/ramda/pull/2098); tests sorta WIP but initial bits passing.

**Comment by tycho01** [comment]
To elaborate, I think the idea of adding the key as the parameter in the case of objects *probably* covers 95% of the use-cases of `R.addIndex`: that is, all cases except where one specifically wanted to use the numerical index when iterating over objects, which seems a bit less common imo. But yeah, I suppose that difference means `addIndex` probably couldn't just be completely ditched in favor of this.
